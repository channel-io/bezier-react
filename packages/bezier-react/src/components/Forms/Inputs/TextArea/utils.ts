import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
