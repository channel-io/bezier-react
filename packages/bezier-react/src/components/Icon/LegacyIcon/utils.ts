/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
