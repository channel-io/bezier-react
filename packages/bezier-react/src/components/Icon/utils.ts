import { type SVGProps } from 'react'

const BEZIER_ICON_SYMBOL = Symbol('bezier.icon')

export type IconSource = (props: SVGProps<SVGSVGElement>) => JSX.Element

export type BezierIcon = IconSource & { [BEZIER_ICON_SYMBOL]: true }

export function isBezierIcon(arg: unknown): arg is BezierIcon {
  return (
    typeof arg === 'function' &&
    arg[BEZIER_ICON_SYMBOL] === true
  )
}

export function createBezierIcon(source: IconSource): BezierIcon {
  source[BEZIER_ICON_SYMBOL] = true
  return source as BezierIcon
}
