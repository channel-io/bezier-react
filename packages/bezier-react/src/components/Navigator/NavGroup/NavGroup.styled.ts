/* Internal dependencies */
import {
  styled,
  css,
} from '~/src/foundation'
import type { InterpolationProps } from '~/src/types/Foundation'

interface WrapperProps extends InterpolationProps {
  active: boolean
}

export const LeftIconWrapper = styled.div`
  display: flex;
  align-items: center;
  min-width: 20px;
  max-width: 20px;
  margin-right: 8px;
`

export const ChevronWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  min-width: 20px;
  max-width: 20px;
  margin-left: 2px;
`

export const RightContentWrapper = styled.div`
  display: flex;
  flex: 1 0 auto;
  align-items: center;
  justify-content: flex-end;
  margin-left: 8px;
`

const activeItemStyle = css`
  color: ${({ foundation }) => foundation?.theme?.['bgtxt-blue-normal']};
  background-color: ${({ foundation }) => foundation?.theme?.['bgtxt-blue-lightest']};
`

const nonActiveItemStyle = css`
  color: ${({ foundation }) => foundation?.theme?.['txt-black-darkest']};

  &:hover {
    background-color: ${({ foundation }) => foundation?.theme?.['bg-black-lighter']};
  }
  
  /*
    ':focus-visible' Pseudo class는 Safari 15.4 이후 지원하기 때문에,
    이를 ':hover' 와 ',' 연산자로 연결할 경우 hover, focus-visible 모두 작동하지 않는 이슈가 있음.
    https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible#browser_compatibility
  */
  &:focus-visible {
    background-color: ${({ foundation }) => foundation?.theme?.['bg-black-lighter']};
  }
`

export const Item = styled.button<WrapperProps>`
  display: flex;
  align-items: center;
  width: 100%;
  height: 28px;
  padding: 0 6px;
  cursor: pointer;
  background-color: transparent;
  border: 0;
  outline: 0;
  ${({ foundation }) => foundation?.rounding?.round6}

  ${({ active }) => (active ? activeItemStyle : nonActiveItemStyle)}

  ${({ interpolation }) => interpolation}
`

export const ChildrenWrapper = styled.ul`
  padding: 0 0 8px;
  margin: 0;
`

export const Wrapper = styled.li`
  display: block;
`
