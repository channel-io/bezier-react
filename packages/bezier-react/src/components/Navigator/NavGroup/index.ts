export { NavGroup } from './NavGroup'
export type { NavGroupProps } from './NavGroup.types'
