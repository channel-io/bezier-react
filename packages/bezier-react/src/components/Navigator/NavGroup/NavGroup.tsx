import React, {
  forwardRef,
  memo,
  useCallback,
} from 'react'

import {
  ChevronSmallDownIcon,
  ChevronSmallRightIcon,
} from '@channel.io/bezier-icons'

import { Typography } from '~/src/foundation'

import { noop } from '~/src/utils/function'
import { isNil } from '~/src/utils/type'

import {
  Icon,
  IconSize,
} from '~/src/components/Icon'
import { Text } from '~/src/components/Text'

import type NavGroupProps from './NavGroup.types'

import {
  ChevronWrapper,
  ChildrenWrapper,
  Item,
  LeftIconWrapper,
  RightContentWrapper,
  Wrapper,
} from './NavGroup.styled'

export const NAV_GROUP_TEST_ID = 'bezier-react-nav-group'
export const NAV_GROUP_LEFT_ICON_TEST_ID = 'bezier-react-nav-group-left-icon'

const NavGroup = forwardRef<HTMLButtonElement, NavGroupProps>(function NavGroup({
  as,
  testId = NAV_GROUP_TEST_ID,
  name,
  style,
  className,
  interpolation,
  children,
  content,
  rightContent,
  leftIcon,
  open,
  active,
  onClick = noop,
  ...rest
}, forwardedRef) {
  const handleClickItem = useCallback((e?: React.MouseEvent) => {
    onClick(e, name)
  }, [
    name,
    onClick,
  ])

  const hasChildren = !isNil(children)
  const chevronIconSource = open ? ChevronSmallDownIcon : ChevronSmallRightIcon
  const ariaName = `${name}Menu`

  return (
    <Wrapper role="none">
      <Item
        {...rest}
        ref={forwardedRef}
        as={as}
        active={active}
        style={style}
        className={className}
        interpolation={interpolation}
        onClick={handleClickItem}
        data-testid={testId}
        role="menuitem"
        aria-haspopup={hasChildren}
        aria-expanded={open}
        aria-controls={ariaName}
      >
        <LeftIconWrapper>
          <Icon
            testId={NAV_GROUP_LEFT_ICON_TEST_ID}
            source={leftIcon}
            size={IconSize.S}
            color={active ? 'bgtxt-blue-normal' : 'txt-black-dark'}
          />
        </LeftIconWrapper>

        <Text typo={Typography.Size14} truncated>
          { content }
        </Text>

        { hasChildren && (
          <ChevronWrapper>
            <Icon
              source={chevronIconSource}
              size={IconSize.S}
              color="txt-black-dark"
            />
          </ChevronWrapper>
        ) }

        { rightContent && (
          <RightContentWrapper>
            { rightContent }
          </RightContentWrapper>
        ) }
      </Item>

      { open && (
        <ChildrenWrapper
          role="menu"
          id={ariaName}
        >
          { open && children }
        </ChildrenWrapper>
      ) }
    </Wrapper>
  )
})

export default memo(NavGroup)
