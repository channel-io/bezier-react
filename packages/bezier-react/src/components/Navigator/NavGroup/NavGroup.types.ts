/* Internal dependencies */
import type {
  BezierComponentProps,
  ChildrenProps,
  ContentProps,
  SideContentProps,
  ActivatableProps,
} from 'Types/ComponentProps'
import { IconName } from 'Components/Icon'

interface NavGroupOptions {
  open?: boolean
  leftIcon: IconName
  name: string
  onClick?: (e?: React.MouseEvent, name?: string) => void
}

export default interface NavGroupProps extends
  BezierComponentProps,
  ChildrenProps,
  ContentProps,
  Pick<SideContentProps, 'rightContent'>,
  Pick<ActivatableProps, 'active'>,
  NavGroupOptions {}
