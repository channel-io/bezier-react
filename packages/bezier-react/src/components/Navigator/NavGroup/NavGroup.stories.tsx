import React from 'react'

import {
  DotIcon,
  SettingsIcon,
} from '@channel.io/bezier-icons'
import {
  type Meta,
  type StoryFn,
  type StoryObj,
} from '@storybook/react'

import {
  Icon,
  IconSize,
} from '~/src/components/Icon'
import { NavItem } from '~/src/components/Navigator/NavItem'

import NavGroup from './NavGroup'
import type NavGroupProps from './NavGroup.types'

const meta: Meta<typeof NavGroup> = {
  component: NavGroup,
}
export default meta

const Template: StoryFn<NavGroupProps> = (args) => (
  <nav
    style={{ width: '240px', padding: '6px' }}
    aria-label="navgroup storybook"
  >
    <ul style={{ margin: '0', padding: '0' }} role="menubar">
      <NavGroup {...args}>
        <NavItem
          name="profile"
          content="채널 프로필"
          href="https://naver.com"
          target="_parent"
        />
        <NavItem
          name="btn"
          content="버튼 설치 및 설정"
          href="https://google.com"
          target="_self"
        />
      </NavGroup>
    </ul>
  </nav>
)

export const Primary: StoryObj<NavGroupProps> = {
  render: Template,

  args: {
    open: true,
    active: false,
    name: 'general',
    content: '일반 설정',
    leftIcon: SettingsIcon,
    rightContent: (
      <Icon source={DotIcon} size={IconSize.XS} color="bgtxt-orange-normal" />
    ),
  },
}
