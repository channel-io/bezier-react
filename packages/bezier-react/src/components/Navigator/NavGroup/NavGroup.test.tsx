import React from 'react'

import { DotIcon } from '@channel.io/bezier-icons'

import { LightTheme } from '~/src/foundation/Colors/Theme'

import { render } from '~/src/utils/testUtils'

import {
  Icon,
  IconSize,
} from '~/src/components/Icon'

import NavGroup, {
  NAV_GROUP_LEFT_ICON_TEST_ID,
  NAV_GROUP_TEST_ID,
} from './NavGroup'
import type NavGroupProps from './NavGroup.types'

describe('NavGroup Test >', () => {
  let props: NavGroupProps

  beforeEach(() => {
    props = {
      leftIcon: 'dot',
      name: 'general',
      content: 'test-content',
      rightContent: <Icon source={DotIcon} size={IconSize.XS} color="bgtxt-orange-normal" />,
    }
  })

  const renderNavItem = (optionProps?: Partial<NavGroupProps>) =>
    render(<NavGroup {...props} {...optionProps} />)

  describe('Snapshot >', () => {
    it('Active', () => {
      const { getByTestId } = renderNavItem({ active: true })

      const rendered = getByTestId(NAV_GROUP_TEST_ID)

      expect(rendered).toMatchSnapshot()
    })

    it('Not active', () => {
      const { getByTestId } = renderNavItem({ active: false })

      const rendered = getByTestId(NAV_GROUP_TEST_ID)

      expect(rendered).toMatchSnapshot()
    })
  })

  describe('LeftIcon Color', () => {
    it('Icon color should be "bgtxt-blue-normal" when active prop is true', () => {
      const { getByTestId } = renderNavItem({ active: true })

      const rendered = getByTestId(NAV_GROUP_LEFT_ICON_TEST_ID)

      expect(rendered).toHaveStyle(`color: ${LightTheme['bgtxt-blue-normal']}`)
    })

    it('Icon color should be "txt-black-dark" when active prop is false', () => {
      const { getByTestId } = renderNavItem({ active: false })

      const rendered = getByTestId(NAV_GROUP_LEFT_ICON_TEST_ID)

      expect(rendered).toHaveStyle(`color: ${LightTheme['txt-black-dark']}`)
    })
  })
})
