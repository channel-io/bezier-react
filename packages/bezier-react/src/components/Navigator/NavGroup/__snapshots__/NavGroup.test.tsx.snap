// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavGroup Test > Snapshot > Active 1`] = `
.c2 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #5E56F0;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c4 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #F4800B;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 20px;
  max-width: 20px;
  margin-right: 8px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-left: 8px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 28px;
  padding: 0 6px;
  cursor: pointer;
  background-color: transparent;
  border: 0;
  outline: 0;
  overflow: hidden;
  border-radius: 6px;
  color: #5E56F0;
  background-color: #5E56F01A;
}

<button
  aria-controls="generalMenu"
  aria-haspopup="false"
  class="c0"
  data-testid="bezier-react-nav-group"
  role="menuitem"
>
  <div
    class="c1"
  >
    <svg
      class="c2"
      color="bgtxt-blue-normal"
      data-testid="bezier-react-nav-group-left-icon"
      fill="none"
      foundation="[object Object]"
      height="20"
      marginbottom="0"
      marginleft="0"
      marginright="0"
      margintop="0"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M17 12a5 5 0 1 1-10 0 5 5 0 0 1 10 0"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="Text typo-14 truncated margin"
    data-testid="bezier-react-text"
  >
    test-content
  </span>
  <div
    class="c3"
  >
    <svg
      class="c4"
      color="bgtxt-orange-normal"
      data-testid="bezier-react-icon"
      fill="none"
      foundation="[object Object]"
      height="16"
      marginbottom="0"
      marginleft="0"
      marginright="0"
      margintop="0"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M17 12a5 5 0 1 1-10 0 5 5 0 0 1 10 0"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</button>
`;

exports[`NavGroup Test > Snapshot > Not active 1`] = `
.c2 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #00000066;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c4 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #F4800B;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 20px;
  max-width: 20px;
  margin-right: 8px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-left: 8px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 28px;
  padding: 0 6px;
  cursor: pointer;
  background-color: transparent;
  border: 0;
  outline: 0;
  overflow: hidden;
  border-radius: 6px;
  color: #000000D9;
}

.c0:hover {
  background-color: #0000000D;
}

.c0:focus-visible {
  background-color: #0000000D;
}

<button
  aria-controls="generalMenu"
  aria-haspopup="false"
  class="c0"
  data-testid="bezier-react-nav-group"
  role="menuitem"
>
  <div
    class="c1"
  >
    <svg
      class="c2"
      color="txt-black-dark"
      data-testid="bezier-react-nav-group-left-icon"
      fill="none"
      foundation="[object Object]"
      height="20"
      marginbottom="0"
      marginleft="0"
      marginright="0"
      margintop="0"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M17 12a5 5 0 1 1-10 0 5 5 0 0 1 10 0"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="Text typo-14 truncated margin"
    data-testid="bezier-react-text"
  >
    test-content
  </span>
  <div
    class="c3"
  >
    <svg
      class="c4"
      color="bgtxt-orange-normal"
      data-testid="bezier-react-icon"
      fill="none"
      foundation="[object Object]"
      height="16"
      marginbottom="0"
      marginleft="0"
      marginright="0"
      margintop="0"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M17 12a5 5 0 1 1-10 0 5 5 0 0 1 10 0"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</button>
`;
